sled-agent/src/nexus.rs

```diff
@@ -39,27 +39,57 @@ pub struct LazyNexusClient {
 }
 
 impl LazyNexusClient {
-    pub fn new(log: Logger, addr: Ipv6Addr) -> Result<Self, ResolveError> {
+    pub fn new_from_subnet(
+        log: Logger,
+        addr: Ipv6Addr,
+    ) -> Result<Self, ResolveError> {
         Ok(Self {
             inner: Arc::new(Inner {
                 log: log.clone(),
                 resolver: Resolver::new_from_ip(
                     log.new(o!("component" => "DnsResolver")),
                     addr,
                 )?,
+                address: Mutex::new(None),
+            }),
+        })
+    }
+
+    pub fn new_from_dns(
+        log: Logger,
+        dns_addrs: Vec<SocketAddr>,
+    ) -> Result<Self, ResolveError> {
+        Ok(Self {
+            inner: Arc::new(Inner {
+                log: log.clone(),
+                resolver: Resolver::new_from_addrs(
+                    log.new(o!("component" => "DnsResolver")),
+                    dns_addrs,
+                )?,
+                address: Mutex::new(None),
             }),
         })
     }
 
     pub async fn get_ip(&self) -> Result<Ipv6Addr, ResolveError> {
-        self.inner.resolver.lookup_ipv6(ServiceName::Nexus).await
+        self.get_addr().await.map(|addr| *addr.ip())
+    }
+
+    pub async fn get_addr(&self) -> Result<SocketAddrV6, ResolveError> {
+        if let Some(addr) = self.inner.address.lock().unwrap().as_ref() {
+            return Ok(*addr);
+        }
+        let addr =
+            self.inner.resolver.lookup_socket_v6(ServiceName::Nexus).await?;
+        *self.inner.address.lock().unwrap() = Some(addr);
+        Ok(addr)
     }
 
     pub async fn get(&self) -> Result<NexusClient, ResolveError> {
-        let address = self.get_ip().await?;
+        let address = self.get_addr().await?;
 
         Ok(NexusClient::new(
-            &format!("http://[{}]:{}", address, NEXUS_INTERNAL_PORT),
+            &format!("http://{address}"),
```

Comment:
This is not a problem right now, but if we ever do internal TLS then this will need to be a DNS name here.  This would be another reason to (eventually) go with the approach of handing reqwest a DNS resolver rather than resolving it ourselves here.

Reply:
Yeah, shifted to supplying reqwest a DNS resolver directly, so this *should* be resolved now. I do still need to specify the port though.